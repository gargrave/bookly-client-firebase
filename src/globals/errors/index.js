// @flow
import { parseError, parseFbError } from './apiErrorParser';
import { validationErrors } from './validationErrors';

export {
  parseError,
  parseFbError,
  validationErrors,
};
