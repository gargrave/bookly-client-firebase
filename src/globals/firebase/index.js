import firebase from 'firebase';
import 'firebase/firestore';

import firebaseConfig from '../../secrets/firebaseConfig.js';

let initialized = false;
let db;

if (!initialized) {
  firebase.initializeApp(firebaseConfig);
  db = firebase.firestore();
  initialized = true;
}

function docToObject(doc) {
  return {
    id: doc.id,
    ...doc.data(),
  };
}

function timestamp() {
  return firebase.firestore.FieldValue.serverTimestamp();
}

export {
  db,
  docToObject,
  timestamp,
};
