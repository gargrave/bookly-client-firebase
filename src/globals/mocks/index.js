import { authorMocks } from './data/authorMocks';
import { bookBucketMocks } from './data/bookBucketMocks';
import { bookMocks } from './data/bookMocks';
import { profileMocks } from './data/profileMocks';
import { userMocks } from './data/userMocks';

export {
  authorMocks,
  bookBucketMocks,
  bookMocks,
  profileMocks,
  userMocks,
};
