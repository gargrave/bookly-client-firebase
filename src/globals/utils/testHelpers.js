import React from 'react';
import { shallow } from 'enzyme';

export class ComponentBuilder {
  constructor(component, defaultProps) {
    this.component = component;
    this.defaultProps = defaultProps;
  }

  shallowGetComponent(extraProps = {}) {
    const props = {
      ...this.defaultProps,
      ...extraProps,
    };
    return shallow(<this.component {...props} />);
  }
}
