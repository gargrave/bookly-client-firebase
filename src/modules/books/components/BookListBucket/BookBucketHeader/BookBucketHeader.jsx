// @flow
import React from 'react'
import { array, bool, func, shape, string } from 'prop-types'
import styled from 'react-emotion'

import type { BookBucket } from '../../../flowtypes'

type Props = {
  bucket: BookBucket,
  expanded: boolean,
  onToggleExpanded: (event: any) => void,
}

export const StyledWrapper = styled('div')`
  align-items: center;
  display: flex;
  justify-content: space-between;
  padding-left: 4px;
  padding-right: 4px;
`

export const StyledAuthorName = styled('span')`
  color: #777;
  font-size: 1.4rem;
  margin-bottom: 2px;
  text-align: left;

  @media only screen and (max-width: 639px) {
    font-size: 1.2rem;
  }
`

export const StyledBookCount = styled('span')`
  font-size: 1rem;
  margin-left: 4px;
`

export const StyledToggle = styled('span')`
  cursor: pointer;
  font-size: 0.9rem;
  margin-left: 8px;
`

const BookBucketHeader = ({ bucket, expanded, onToggleExpanded }: Props) => {
  return (
    <StyledWrapper>
      <StyledAuthorName>
        {bucket.author}
        <StyledBookCount>{`(${bucket.books.length})`}</StyledBookCount>
      </StyledAuthorName>
      <StyledToggle onClick={onToggleExpanded}>
        {expanded ? 'collapse' : 'expand'}
      </StyledToggle>
    </StyledWrapper>
  )
}

BookBucketHeader.propTypes = {
  bucket: shape({
    author: string,
    books: array,
  }).isRequired,
  expanded: bool.isRequired,
  onToggleExpanded: func.isRequired,
}

export default BookBucketHeader
