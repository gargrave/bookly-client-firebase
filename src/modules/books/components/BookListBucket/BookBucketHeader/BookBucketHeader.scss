.bookBucketHeader {
  color: #777;
  font-size: 1.4rem;
  margin-bottom: 2px;
  text-align: left;
}

@media only screen and (max-width: 639px) {
  .bookBucketHeader {
    font-size: 1.2rem;
  }
}
