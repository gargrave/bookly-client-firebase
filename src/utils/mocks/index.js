import authorMocks from './data/authors';
import userMocks from './data/users';

export {
  authorMocks,
  userMocks,
};
