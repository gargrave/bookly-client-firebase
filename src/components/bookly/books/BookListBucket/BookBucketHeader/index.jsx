// @flow
import React from 'react';
import { array, shape, string } from 'prop-types';

import type { BookBucket } from '../../../../../constants/flowtypes/';

import { buildClasses } from '../../../../../utils/cssHelpers';

type Props = {
  bucket: BookBucket,
};

function BookBucketHeader({
  bucket,
}: Props) {
  return (
    <div className={buildClasses(['book-bucket__header'])}>
      <p>{bucket.author}</p>
    </div>
  );
}

BookBucketHeader.propTypes = {
  bucket: shape({
    author: string,
    books: array,
  }).isRequired,
};

export default BookBucketHeader;
