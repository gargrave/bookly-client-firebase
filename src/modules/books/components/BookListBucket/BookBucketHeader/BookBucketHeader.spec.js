import { bookBucketMocks } from '../../../../../globals/mocks/'
import { ComponentBuilder } from '../../../../../utils/testHelpers'

import BookBucketHeader, {
  StyledAuthorName,
  StyledToggle,
  StyledWrapper,
} from './BookBucketHeader'

const defaultProps = {
  bucket: bookBucketMocks[0],
  expanded: false,
  onToggleExpanded: jest.fn(),
}

const builder = new ComponentBuilder(BookBucketHeader, defaultProps)

describe('BookBucketHeader', () => {
  let component

  describe('basic rendering', () => {
    it('matches the snapshot', () => {
      component = builder.shallowGetComponent()
      expect(component).toMatchSnapshot()
    })

    it('renders correctly', () => {
      component = builder.mountComponent()
      // const authorString = `${bookBucketMocks[0].author}`
      // expect(component.find({ children: authorString })).toHaveLength(1)

      const wrapperNode = component.find(StyledWrapper)
      expect(wrapperNode).toHaveLength(1)

      const authorNode = component.find(StyledAuthorName)
      expect(authorNode).toHaveLength(1)
      let authorRE = new RegExp(defaultProps.bucket.author)
      expect(authorRE.test(authorNode.text())).toBe(true)

      const expandToggleNode = component.find(StyledToggle)
      expect(expandToggleNode.length).toBe(1)
    })
  })

  describe('expand/collapse', () => {
    it('shows "expand" when collapsed', () => {
      component = builder.mountComponent()
      expect(component.find(StyledToggle).text()).toBe('expand')
    })

    it('shows "expand" when collapsed', () => {
      component = builder.mountComponent({ expanded: true })
      expect(component.find(StyledToggle).text()).toBe('collapse')
    })

    it('calls the callback when clicked', () => {
      component = builder.mountComponent()
      expect(defaultProps.onToggleExpanded).toHaveBeenCalledTimes(0)
      component.find(StyledToggle).simulate('click')
      expect(defaultProps.onToggleExpanded).toHaveBeenCalledTimes(1)
    })
  })
})
