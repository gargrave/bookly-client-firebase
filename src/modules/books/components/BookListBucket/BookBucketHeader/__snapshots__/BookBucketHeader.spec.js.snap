// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookBucketHeader basic rendering matches the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <BookBucketHeader
    bucket={
      Object {
        "author": "Aldous Huxley",
        "books": Array [
          Object {
            "author": Object {
              "bookCount": 1,
              "created": 2017-07-01T01:16:43.881Z,
              "firstName": "Aldous",
              "id": "oiw223h3qndlksadffl3ksdf",
              "lastName": "Huxley",
              "updated": 2017-07-01T01:16:43.881Z,
            },
            "created": 2018-01-14T01:36:26.220Z,
            "id": "sv3hmbefpc95MRXQsTHr",
            "sortBy": "",
            "title": "Brave New World",
            "updated": 2018-01-14T01:36:26.220Z,
          },
          Object {
            "author": Object {
              "bookCount": 1,
              "created": 2017-07-01T01:16:43.881Z,
              "firstName": "Aldous",
              "id": "oiw223h3qndlksadffl3ksdf",
              "lastName": "Huxley",
              "updated": 2017-07-01T01:16:43.881Z,
            },
            "created": 2018-01-15T01:36:26.220Z,
            "id": "Hmh9eR3rs5bMvsTXfpcQ",
            "sortBy": "",
            "title": "Island",
            "updated": 2018-01-15T01:36:26.220Z,
          },
          Object {
            "author": Object {
              "bookCount": 1,
              "created": 2017-07-01T01:16:43.881Z,
              "firstName": "Aldous",
              "id": "oiw223h3qndlksadffl3ksdf",
              "lastName": "Huxley",
              "updated": 2017-07-01T01:16:43.881Z,
            },
            "created": 2018-01-16T01:36:26.220Z,
            "id": "eRpcQ5H3rsTXfmh9bMvs",
            "sortBy": "",
            "title": "The Doors of Perception",
            "updated": 2018-01-16T01:36:26.220Z,
          },
        ],
      }
    }
    expanded={false}
    onToggleExpanded={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <Styled(span)>
          Aldous Huxley
          <Styled(span)>
            (3)
          </Styled(span)>
        </Styled(span)>,
        <Styled(span)
          onClick={[Function]}
        >
          expand
        </Styled(span)>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            "Aldous Huxley",
            <Styled(span)>
              (3)
            </Styled(span)>,
          ],
        },
        "ref": null,
        "rendered": Array [
          "Aldous Huxley",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "(3)",
            },
            "ref": null,
            "rendered": "(3)",
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "expand",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "expand",
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <Styled(span)>
            Aldous Huxley
            <Styled(span)>
              (3)
            </Styled(span)>
          </Styled(span)>,
          <Styled(span)
            onClick={[Function]}
          >
            expand
          </Styled(span)>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              "Aldous Huxley",
              <Styled(span)>
                (3)
              </Styled(span)>,
            ],
          },
          "ref": null,
          "rendered": Array [
            "Aldous Huxley",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "(3)",
              },
              "ref": null,
              "rendered": "(3)",
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "expand",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "expand",
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
